File: mcp-code-intelligence-nodejs/src/orchestration/local/manager.ts
/**
 * Local server manager — manages multiple child MCP server processes.
 * Behavioral parity with Kotlin LocalServerManager.kt.
 */

import { OrchestrationConfig, ServerEntry, enabledServers } from '../config.js';
import { ServerProcess, ServerState } from './process.js';

export class LocalServerManager {
  private config: OrchestrationConfig;
  private servers = new Map<string, ServerProcess>();
  private healthInterval: ReturnType<typeof setInterval> | null = null;

  constructor(config: OrchestrationConfig) { this.config = config; }

  updateConfig(newConfig: OrchestrationConfig): void { this.config = newConfig; }

  /** Start all enabled servers. Returns count of successfully started. */
  async startAll(): Promise<number> {
    const entries = enabledServers(this.config);
    console.error(`[orchestration] Starting ${entries.size} child servers...`);
    let started = 0;
    for (const [name, entry] of entries) {
      const server = new ServerProcess(name, entry);
      this.servers.set(name, server);
      if (await server.start()) started++;
      else console.error(`[${name}] Failed to start`);
    }
    this.startHealthMonitor();
    return started;
  }

  /** Stop all child servers gracefully. */
  stopAll(): void {
    if (this.healthInterval) { clearInterval(this.healthInterval); this.healthInterval = null; }
    for (const server of this.servers.values()) server.stop();
    this.servers.clear();
    console.error('[orchestration] All child servers stopped');
  }

  /** Call a tool on the specified server. */
  async callTool(serverName: string, toolName: string, args: Record<string, any>, timeoutMs: number): Promise<any> {
    const server = this.servers.get(serverName);
    if (!server) throw new Error(`Server '${serverName}' not found (tool: '${toolName}')`);
    if (server.state !== ServerState.ACTIVE) throw new Error(`Server '${serverName}' is ${server.state}`);
    return server.callTool(toolName, args, timeoutMs);
  }

  /** Find which server owns a given tool name. */
  findServerForTool(toolName: string): string | null {
    for (const [name, server] of this.servers) {
      if (server.state !== ServerState.ACTIVE) continue;
      if (server.tools.some((t) => t.name === toolName)) return name;
    }
    return null;
  }

  /** Get all tools from all active child servers. */
  getAllTools(): Array<[string, Record<string, any>]> {
    const result: Array<[string, Record<string, any>]> = [];
    for (const [name, server] of this.servers) {
      if (server.state === ServerState.ACTIVE) {
        for (const tool of server.tools) result.push([name, tool]);
      }
    }
    return result;
  }

  /** Get status of all managed servers. */
  getStatus(): Map<string, ServerState> {
    const result = new Map<string, ServerState>();
    for (const [name, server] of this.servers) result.set(name, server.state);
    return result;
  }

  /** Get detailed status info. */
  getServerStatusInfo(): Array<{ name: string; state: string; toolCount: number }> {
    return [...this.servers.entries()].map(([name, s]) => ({
      name, state: s.state, toolCount: s.tools.length,
    }));
  }

  private startHealthMonitor(): void {
    const intervalMs = this.config.settings.healthCheckIntervalMs;
    this.healthInterval = setInterval(() => this.checkHealth(), intervalMs);
  }

  /** Retry starting servers that are in FAILED state. Returns names of recovered servers. */
  async retryFailedServers(): Promise<string[]> {
    const recovered: string[] = [];
    for (const [name, server] of this.servers) {
      if (server.state !== ServerState.FAILED) continue;
      console.error(`[${name}] Retrying failed server...`);
      if (await server.start()) {
        recovered.push(name);
        console.error(`[${name}] Recovered — now active with ${server.tools.length} tools`);
      } else {
        console.error(`[${name}] Still failing`);
      }
    }
    return recovered;
  }

  private async checkHealth(): Promise<void> {
    for (const [name, server] of this.servers) {
      if (server.state === ServerState.FAILED) {
        console.error(`[${name}] Health check: retrying failed server`);
        if (await server.start()) {
          console.error(`[${name}] Recovered via health check`);
        }
        continue;
      }
      if (server.state !== ServerState.ACTIVE) continue;
      if (!server.isAlive() || !await server.healthCheck()) {
        console.error(`[${name}] Unhealthy — attempting restart`);
        const maxRetries = this.config.settings.maxRestartRetries;
        if (!await server.restart(maxRetries)) {
          console.error(`[${name}] Permanently dead after ${maxRetries} retries`);
        }
      }
    }
  }
}

File: mcp-code-intelligence-python/src/mcp_code_intel/orchestration/local/manager.py
"""Local server manager — manages multiple child MCP server processes.

Start, stop, health monitoring, tool routing.
Behavioral parity with Kotlin LocalServerManager.kt.
"""

from __future__ import annotations

import asyncio
import sys
from typing import Any

from ..config import OrchestrationConfig, ServerEntry, TransportType
from .process import ServerProcess, ServerState
from .transport import detect_transport
from .http_stream_process import HttpStreamProcess


class LocalServerManager:
    """Manages multiple child MCP server processes."""

    def __init__(self, config: OrchestrationConfig) -> None:
        self._config = config
        self._servers: dict[str, ServerProcess | HttpStreamProcess] = {}
        self._health_task: asyncio.Task | None = None

    def update_config(self, new_config: OrchestrationConfig) -> None:
        """Update config reference (used by hot-reload)."""
        self._config = new_config

    async def start_all(self) -> int:
        """Start all enabled servers. Returns count of successfully started."""
        entries = self._config.enabled_servers()
        _log(f"Starting {len(entries)} child servers...")
        started = 0
        for name, entry in entries.items():
            transport = detect_transport(entry)
            if transport == TransportType.HTTP_STREAM:
                server = HttpStreamProcess(name, entry)
            else:
                server = ServerProcess(name, entry)
            self._servers[name] = server
            if await server.start():
                started += 1
            else:
                _log(f"[{name}] Failed to start (transport: {transport.value})")
        self._start_health_monitor()
        return started

    def stop_all(self) -> None:
        """Stop all child servers gracefully."""
        if self._health_task:
            self._health_task.cancel()
            self._health_task = None
        for server in self._servers.values():
            server.stop()
        self._servers.clear()
        _log("All child servers stopped")

    async def call_tool(self, server_name: str, tool_name: str, args: dict, timeout_ms: int) -> Any:
        """Call a tool on the specified server."""
        server = self._servers.get(server_name)
        if not server:
            raise RuntimeError(f"Server '{server_name}' not found (tool: '{tool_name}')")
        if server.state != ServerState.ACTIVE:
            raise RuntimeError(f"Server '{server_name}' is {server.state.value}")
        return await server.call_tool(tool_name, args, timeout_ms)

    def find_server_for_tool(self, tool_name: str) -> str | None:
        """Find which server owns a given tool name."""
        for name, server in self._servers.items():
            if server.state != ServerState.ACTIVE:
                continue
            if any(t.get("name") == tool_name for t in server.tools):
                return name
        return None

    def get_all_tools(self) -> list[tuple[str, dict]]:
        """Get all tools from all active child servers."""
        result = []
        for name, server in self._servers.items():
            if server.state == ServerState.ACTIVE:
                result.extend((name, t) for t in server.tools)
        return result

    def get_status(self) -> dict[str, ServerState]:
        """Get status of all managed servers."""
        return {name: s.state for name, s in self._servers.items()}

    def get_server_status_info(self) -> list[dict]:
        """Get detailed status info for orchestration_status tool."""
        return [
            {"name": name, "state": s.state.value, "toolCount": len(s.tools)}
            for name, s in self._servers.items()
        ]

    async def retry_failed_servers(self) -> list[str]:
        """Retry starting servers that are in FAILED state. Returns names of newly active servers."""
        recovered: list[str] = []
        for name, server in list(self._servers.items()):
            if server.state != ServerState.FAILED:
                continue
            _log(f"[{name}] Retrying failed server...")
            server.retry_count = 0  # Reset retry count for fresh attempt
            if await server.start():
                recovered.append(name)
                _log(f"[{name}] Recovered — now active with {len(server.tools)} tools")
            else:
                _log(f"[{name}] Still failing")
        return recovered

    def _start_health_monitor(self) -> None:
        """Start background health check task."""
        interval_s = self._config.settings.health_check_interval_ms / 1000.0
        self._health_task = asyncio.create_task(self._health_loop(interval_s))

    async def _health_loop(self, interval_s: float) -> None:
        """Periodically check health of all active servers."""
        try:
            while True:
                await asyncio.sleep(interval_s)
                await self._check_health()
        except asyncio.CancelledError:
            pass

    async def _check_health(self) -> None:
        """Check each active server, restart if unhealthy. Also retry FAILED servers."""
        for server in list(self._servers.values()):
            if server.state == ServerState.FAILED:
                _log(f"[{server.name}] Health check: retrying failed server")
                server.retry_count = 0
                if await server.start():
                    _log(f"[{server.name}] Recovered via health check")
                continue
            if server.state != ServerState.ACTIVE:
                continue
            if not server.is_alive() or not await server.health_check():
                _log(f"[{server.name}] Unhealthy — attempting restart")
                max_retries = self._config.settings.max_restart_retries
                if not await server.restart(max_retries):
                    _log(f"[{server.name}] Permanently dead after {max_retries} retries")


def _log(msg: str) -> None:
    print(f"[orchestration] {msg}", file=sys.stderr, flush=True)
